### PR TITLE
Fix circular dependencies

### PR DESCRIPTION
**Pull request by eatonphil**
Circular dependencies left me with a nasty bug I discovered in #536 (why it's taken me so long to get tests passing in that PR). It took forever to figure out that this actually was a circular dependency issue. Modules were seemingly imported multiple times and certain exported values within the module were undefined. I refactored the api code a bit to not need circular dependencies any more.

While I was trying to figure out what was going on, I also upgraded all packages in case there was an issue in babel or sth. I upgraded the linter during that time and started fixing linter issues - so that's why there are a few random changes. I'll be submitting another PR soon with all packages upgraded and the full linter changes (some 300+ changes or sth).

This PR is a WIP because I haven't yet tested out how the code works for real - I just finally got tests passing. However, I recommend this PR get merged as soon as possible because the bugs I saw in #536 could show up for anyone at any time because there are circular dependencies in our code _right now_.

Also, there's some weird stuff going on locally where this test keeps failing:

```
) renders the projected time of the first backup when its window is < $NOW
     linodes/components/Linode renderBackupStatus
12 10 2016 22:31:00.533:WARN [reporter]: SourceMap position not found for trace: undefined
     expected 'In ~an hour' to equal 'In ~a day'
AssertionError@webpack:///~/assertion-error/index.js:74:0 <- test/index.js:303:25
assert@webpack:///~/chai/lib/chai/assertion.js:107:0 <- test/index.js:4635:32
assertEqual@webpack:///~/chai/lib/chai/core/assertions.js:487:0 <- test/index.js:5152:19
webpack:///~/chai/lib/chai/utils/addMethod.js:41:0 <- test/index.js:4222:31
test/index.js:104379:9517
```

I do not understand - it even fails on master (with no changes) locally here. @na3d @abemassry @briansteffens Can anyone corroborate?

### CODE REVIEW

**Comment by eatonphil** [comment]
Things look good to me from a brief run-though of the actual site.

File: src/api/gen.js [hunk 1]
```diff
@@ -338,7 +148,7 @@ export function genReducer(_config) {
     const { ids } = action;
 
     let name = null;
-    for (let i = 0; i < names.length; i++) {
+    for (let i = 0; i < names.length; i += 1) {
```
Comment (abemassry) [inline src/api/gen.js]: to prevent implicit type coercion

File: src/api/gen.js [hunk 2]
```diff
@@ -351,7 +161,7 @@ export function genReducer(_config) {
     const keys = Object.keys(config.subresources);
     let subkey = null;
     let subconfig = null;
-    for (let i = 0; i < keys.length; i++) {
+    for (let i = 0; i < keys.length; i += 1) {
```
Comment (abemassry) [inline src/api/gen.js]: to prevent implicit type coercion

**Comment by eatonphil** [comment]
Thanks!

**Pull request merged**